liferay/clay
# fix(@clayui/css): remove `node-sass` and add new build file

:warning: WARNING, this is a draft pull request :warning:

---
 - This commit message will be edited before the merge but I thought it
   would be a good idea to add as much information as possible.
 - The CI (GitHub actions) will probably fail
 - There might be conflicts, but I'll rebase with `master` later on
 - There's a lot to read, sorry about that
---

This is a "first step" to fix the [Modernize the clay-css package build system](#4132)
issue: in reality, this is more about removing `node-sass` from clay, and
along the way remove other dependencies like `gulp` or `metalsmith`, as well
as updating other dependencies like `gatsby`. All that to say that there's
still a **lot** of work expected to make clay work on different node
versions, because `clayui.com` also depends on `node-sass`, as well
as the `clay-charts` package.

In this change we change `node-sass`, with `sass` in order to be able to
- Compile `clay-css` with node 10, lts (and hopefully latest soon)
- Remove `gulp` tasks
- Remove other unused dependencies

**Test plan**
- Fetch this branch
- Run `yarn`
- Run `yarn workspace @clayui/css build`
- Assert that the `clay-css` is still working as expected (i.e. nothing broke)
- Check that the `icons.svg` file is correct
-  All this should work with node 10.x and node lts (14.x right now).

Be **aware** that even if the above works, we still have work to do here,
because `clayui.com` and `clay-charts`, also depend on `node-sass`, so we'll
have to do something about it. By switching to `sass` instead of `node-sass`
in our workspace, in this branch the `clayui.com` project doesn't work as
expected, and `clay-charts` throws TypeScript [errors](https://gist.github.com/julien/2311f7d05e6e2a1aba5d87257c75f678).

Things left to do:

- [ ] Clean up remaining "unused" dependecies in `clay-css`
- [ ] Update `clayui.com` directory
- [ ] Discuss (see below)

Thing I'd like to discuss:

- When building `clay-css`, on the `master` branch, the `lib` directory
contains an `css/cadmin` directory, with two child directories, but they are empty.
I guess this isn't expected, but I wanted to make sure.

- This change, is a replacement for the `gulp compile` task, but not 
for the `gulp build` task. I assume the `gulp build` task is there for 
our "test site", but if someone could confirm, it would great.

- In the `gulp` based build:
     - We copy existing scss files from the `src/scss` directory to the `lib/css` directory
    - We compile scss file from the `lib/css` directory (to the `lib/css` directory)
    - We then remove/delete the `*.scss` files copied previously in the `lib/css` directory

     So my **question is**: couldn't we just compile the scss files from the 
     `src/scss` directory to the `lib/css` directory to avoid copying  
     and removing? If it's not the case, I'd like to understand why.

- About the `icons.svg` file that we generate, @bryceosterhaus mentioned 
[something](https://gist.github.com/julien/d419b03c42a4864fef6f58dbbcd40411#gistcomment-3908108) about the `viewBox` attribute. If we could check that the "new" 
`icons.svg` file is the same as the "old" one, it would be great.

- When installing the `@clayui/css` package from npm, we have a `src` and `lib`
directory. The files in the `src/images/icons` directory are exactly the same
ones as in the `lib/css/images/icons` directory, the only difference is that `lib/css` also
contains the `icons.svg` file that we generate. Is this expected?

- Same thing as above but for the `src/js` and `lib/js` directories.

- As part of the "migration" from `sass` to `node-sass`, we'll also have to
update the `README.md` file in `packages/clay-css` because it mentions
`gulp`

- Nothing to do with this change, but running `yarn format` changed a lot
of untouched files (which I obviously didn't commit) but I thought I'd mention
it since it doesn't seem normal.

- Do we know if the `sassdoc` task works with `sass`?

- For some reason with node 14 (lts at the time of writing this), when running
`yarn lint` from the root of the repository we get [errors](https://gist.github.com/julien/a908a49786e06c16d283998160b96cd4).

This is very interesting task, and it reminded me of this [video](https://www.youtube.com/watch?v=lKXe3HUG2l4).

It's a very good exercise to realize that maintaining node projects, isn't easy.
